### PR TITLE
Refactor cross chain transfer.

### PR DESCRIPTION
**Pull request by shiwk**

### CODE REVIEW

File: contract/AElf.Contracts.MultiToken/TokenContract_Actions.cs [hunk 1]
```diff
@@ -110,18 +129,70 @@ public override Empty Transfer(TransferInput input)
...
+            return new Empty();
+        }
+
         /// <summary>
-        /// Transfer token form a chain to another chain
+        /// Transfer token form this chain to another one.
         /// </summary>
         /// <param name="input"></param>
         /// <returns></returns>
         public override Empty CrossChainTransfer(CrossChainTransferInput input)
```
Comment (aerkol) [inline contract/AElf.Contracts.MultiToken/TokenContract_Actions.cs]: When doing cross chain transfer, we reduce `Supply` in current chain by burning. When issuing more tokens, we check whether `Supply` is equal or lower than `TotalSupply`. In this case, how do we prevent issuer to issue without exceeding `TotalSupply`? (In a scenario where issuer issues new tokens then do a cross chain transfer and repeat doing this) Maybe we introduce one more variable `TotalIssued` to track the number of tokens issued?

`Supply`: Number of tokens in current chain (-/+ when doing cross chain transfer)
`TotalIssued`: Total number of tokens issued until now (+ every time new tokens issued on parent chain)
`TotalSupply`: Maximum number of tokens that can be issued

File: contract/AElf.Contracts.MultiToken/TokenContract_Helper.cs [hunk 2]
```diff
@@ -53,14 +59,75 @@ private void AssertLockAddress(string symbol)
...
+        private void CrossChainVerify(Hash transactionId, long parentChainHeight, int chainId, IEnumerable<Hash> merklePath)
+        {
+            var verificationInput = new VerifyTransactionInput
+            {
+                TransactionId = transactionId,
+                ParentChainHeight = parentChainHeight,
+                VerifiedChainId = chainId
+            };
+            verificationInput.Path.AddRange(merklePath);
+            var verificationResult = GetValidCrossChainContractReferenceState().VerifyTransaction.Call(verificationInput);
```
[the remark this answers is not in the document]
Comment (shiwk) [reply contract/AElf.Contracts.MultiToken/TokenContract_Helper.cs]: @aerkol The thing is receiver chain knows nothing about transfer chain, txn count, block size, or whatever things. This is how merkle proof works. We can use other ways to fix this, like txn fees. BTW, txn size is indeed a DOS concerning and maybe txn fee can be set considering with txn size.

File: contract/AElf.Contracts.MultiToken/TokenContract_Actions.cs (hunk 1 above)
Comment (shiwk) [reply contract/AElf.Contracts.MultiToken/TokenContract_Actions.cs]: @aerkol Yeah, we also noticed this issue and it is to be solved.

File: contract/AElf.Contracts.MultiToken/TokenContract_Helper.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (aerkol) [reply contract/AElf.Contracts.MultiToken/TokenContract_Helper.cs]: Receiver chain knows the merkle root of the source block that the token burn happened, what I meant to say was whether we could store tree depth together with merkle root. Then receiver chain can get tree depth the same way it gets merkle root then compare it with the merkle path from input.

[the remark this answers is not in the document]
Comment (shiwk) [reply contract/AElf.Contracts.MultiToken/TokenContract_Helper.cs]: @aerkol  I see what you mean but basically it is impossible. What receiver chain gets is the root calculated in parent chain with multi chains data(imagine a big tree). Actually, for cross chain module, receiver chain doesn't know other chains except parent chain. As a result, it only knows the root result of this big tree and no idea about how many cousin chains involved in this tree, which is too far from the txn count in one block from a specific chain.

[the remark this answers is not in the document]
Comment (aerkol) [reply contract/AElf.Contracts.MultiToken/TokenContract_Helper.cs]: Well, then I need to check the current indexing mechanism again...

**Pull request merged**